Introduce thenAsync method on Future

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58753 Fix race condition in TP agent
* #58675 Make Future always extract DataPtrs in RPC tests
* #56863 Ensure async_execution works with CUDAFuture
* #57355 Avoid re-doing CUDA stream sync in OwnerRRef
* #58428 Make TP agent use streams from Future when sending response
* #58674 Set and propagate devices in RRef completion future
* #58427 Set streams when invoking UDFs
* #58426 Create CUDA-aware futures in RequestCallback
* #58425 Provide pre-extracted DataPtrs when completing a Future with a Message
* #58424 Allow Future::then to return pre-extracted DataPtrs
* #58423 Always use intrusive_ptr for Message (2 out of 2)
* #58422 Always use intrusive_ptr for Message (1 out of 2)
* #58421 Prevent using anything other than intrusive_ptr for Future
* #58420 Migrate remaining shared_ptr<Future> to intrusive_ptr
* #57861 Make remaining autograd methods return futures
* #57860 Make remaining RRef methods return futures
* #57859 Unify fetching RRefs
* #57858 Deduplicate Python object serialization
* #57857 Simplify process(Script|Python)(Remote)?Call
* #57856 Unify assignment of OwnerRRef result
* #57855 Make processPythonExecution return a future
* #57854 Remove getScriptRemoteCallType
* #57853 Simplify OwnerRRef completion
* #57852 Unify async execution for JIT functions
* #57851 Unify invoking JIT functions
* #57850 Unify invoking JIT operands
* #57849 Make some methods of RequestCallback return void instead of bool
* #57848 Centralize setting messageId in RequestCallback
* #57847 Make RequestCallback collect Futures from methods, rather than providing them
* #57846 Add helpers to manipulate futures
* #57638 Reduce overhead when Future invokes callbacks inline
* **#57637 Introduce thenAsync method on Future**

I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

lw committed:
Introduce thenAsync method on Future

I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -494,6 +494,31 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
     return childFut;
   }
 
+  c10::intrusive_ptr<Future> thenAsync(
```
Looks like there is no usage of this API in this stack. Do we plan to use it in the near future?

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -506,6 +506,31 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
+            childFut->setError(std::current_exception());
+            return;
+          }
+          intermediateFut->addCallback([childFut](Future& intermediateFut) {
```
move childFut?

mrshenli reviewed (approved):
LGTM, assuming this will be tested in commits on top

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]

lw committed:
Update on "Introduce thenAsync method on Future"


I had proposes a similar method in https://github.com/pytorch/pytorch/pull/48790, although that PR was exposing it to Python and thus requires a bit more work. This PR only introduces this method as a C++ API. Python can be added later.

This new method is useful when one wants to use `then` but the callback does perform some async operation itself, and one wants to "reconcile" the future produced inside the callback with the one produced outside.

Differential Revision: [D28222809](https://our.internmc.facebook.com/intern/diff/D28222809/)

[ghstack-poisoned]
